### PR TITLE
[REVIEW] Change `type_id` to enum class

### LINKED ISSUE #5384: [FEA] Make `type_id` an enum class
**Is your feature request related to a problem? Please describe.**

The [`type_id` ](https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/include/cudf/types.hpp#L180) enum is a "plain" enum as opposed to a scoped enum class. This was added before we knew that enum classes could be supported from Cython, and was never changed to an `enum class` after we confirmed `enum class` can be supported from Cython. 

Making it an `enum class` may have prevented issues like what happened in https://github.com/rapidsai/cudf/pull/5383#pullrequestreview-424584038 

**Describe the solution you'd like**

Make `type_id` an `enum class`.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
closes #5384

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/transform/transform.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, NVIDIA CORPORATION.
+m* Copyright (c) 2019, NVIDIA CORPORATION.
```
Comment (cwharris) [inline cpp/src/transform/transform.cpp:2]:
```suggestion
 * Copyright (c) 2019, NVIDIA CORPORATION.
```

Comment (rgsl888prabhu) [reply cpp/src/transform/transform.cpp:2]: Yes, and that's why the style check is passing, but will fix this one

**Comment by rgsl888prabhu** [comment]
> The formatting seems to be incorrect for many of the cpp changes. Was clang format run on this with the latest styles?

Yes, and that's why the style check is passing.

File: python/cudf/cudf/_lib/transform.pyx [hunk 2]
```diff
@@ -79,10 +80,14 @@ def transform(Column input, op):
     compiled_op = cudautils.compile_udf(op, nb_signature)
     c_str = compiled_op[0].encode('UTF-8')
     np_dtype = np.dtype(compiled_op[1])
+    print("RGSL  :np_dtype is ", np_dtype)
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/transform.pyx:83]: Stale debug code?
```suggestion
```

**Pull request merged**